kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/SignalManager.java

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright 2015 Cloudera Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.kitesdk.data.spi.filesystem;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.kitesdk.data.DatasetException;
+import org.kitesdk.data.DatasetIOException;
+import org.kitesdk.data.spi.Constraints;
+
+import com.google.common.base.Joiner;
+
+/**
+ * Manager for creating, and checking {@link View#isReady() ready} signals.
+ * Stored in a filesystem, typically HDFS.
+ */
+public class SignalManager {
+
+  private final Path signalDirectory;
+  private final FileSystem rootFileSystem;
+
+  private static final String UNBOUNDED_CONSTRAINT = "unbounded";
+
+  /**
+   * Creates a new signal manager using the given signal directory.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager
+   *                        stores signals.
+   *
+   * @return a signal manager instance.
+   */
+  public static SignalManager create(Configuration conf, Path signalDirectory) {
+    try {
+      FileSystem rootFileSystem = signalDirectory.getFileSystem(conf);
+
+      rootFileSystem.mkdirs(signalDirectory);
+
+      return new SignalManager(signalDirectory, rootFileSystem);
+    } catch (IOException e) {
+      throw new DatasetIOException("Unable to create signal manager directory: "
+              + signalDirectory, e);
+    }
+  }
+
+  /**
+   * Loads a signal manager that stores metadata data under the given signal
+   * metadata directory if it exists. Returns <code>null</code> if it does not.
+   *
+   * @param conf the Hadoop configuration
+   * @param signalDirectory directory in which the manager stores signals.
+   *
+   * @return a signal manager instance, or <code>null</code> if the given
+   * directory does not exist.
+   */
+  public static SignalManager load(Configuration conf, Path signalDirectory) {
```

Comment:
I think it would be a cleaner API to always return a SignalManager, and possibly just expose the constructor instead of create and load. Then the manager can handle whether or not the directory exists internally and internal libraries can use a null manager to indicate that the underlying Dataset implementation doesn't support a signal manager. That would help solve the casting problem I noted.